[PROTOTYPE] two pass scan: no_init processing for inclusive scan

danhoeflinger opened the pull request:
Adding no-init inclusive scan.  Initial draft, can probably clean this up a bit, but this is passing tests.  

Using existing `__init_processing` where possible, but we need to reach in to handle some cases directly within the kernel to avoid default construction.  I attempt to keep order of operations consistent here with the existing code, skipping calculations which don't exist.

Refactoring repeating code into scan_through_elements_helper with some constexpr branches to represent required differences.

Also adding test to exercise new API.

danhoeflinger commented:
I've selectively applied the clang formatting changes so that I don't dilute the diff too much for this PR. 
Once we can merge #1662 and this PR, I'll just push a cleanup clang format commit directly to the prototype branch to get the whole file with a clean clang format.

mmichel11 commented on include/oneapi/dpl/experimental/kt/two_pass_scan.h:
```diff
@@ -149,15 +149,115 @@ sub_group_scan_partial(const SubGroup& sub_group, ValueType& value, BinaryOp bin
+scan_through_elements_helper(const SubGroup& sub_group, UnaryOp unary_op, BinaryOp binary_op,
+                             LazyValueType& sub_group_carry, InRange __in_rng, OutRange __out_rng,
+                             std::size_t start_idx, std::size_t M, std::size_t J, std::size_t subgroup_start_idx,
+                             std::size_t sub_group_id, std::size_t active_subgroups)
```
`sub_group_id` can be `std::uint32_t` as that is the return type from the query in the SYCL spec.

It may be worth looking into if using smaller sized types shows any benefit for some of these variables where it is applicable but that is larger task beyond here this PR.

danhoeflinger replied:
Its true, we can probably shrink a lot of these types down.  I can try to see what low hanging fruit there is here.

danhoeflinger replied:
I made some improvements to the types.

danhoeflinger commented on include/oneapi/dpl/experimental/kt/two_pass_scan.h:
```diff
@@ -230,68 +329,18 @@ two_pass_scan(sycl::queue q, _InRng&& __in_rng, _OutRng&& __out_rng,
+                    scan_through_elements_helper<VL, IsInclusive,
+                                                 /*InitPresent=*/false,
+                                                 /*CaptureOutput=*/false, J_max>(
+                        sub_group, unary_op, binary_op, sub_group_carry, __in_rng, nullptr, start_idx, M, J,
```
I removed the `require` for `__out_rng` for this kernel, because it is unused in this kernel.  We would need to re-introduce that to (not) use it here.  I thought about just using `__in_rng`, but it seems that some empty type is better.  

Also, this `nullptr` attempts to make it clear that we will not be using this parameter or capturing anything to the output.

Would you prefer `std::void_t` perhaps?

mmichel11 replied:
Ah, I forgot it's unused in this kernel. I don't think it is worth it to bring `__out_rng` in here then. I am fine with leaving it with `nullptr`. It is clear what that means to me.

danhoeflinger commented on include/oneapi/dpl/experimental/kt/two_pass_scan.h:
```diff
@@ -149,15 +149,115 @@ sub_group_scan_partial(const SubGroup& sub_group, ValueType& value, BinaryOp bin
 {
     using namespace sycl;
-
+    using ValueType = oneapi::dpl::__internal::__value_t<_InRng>;
```
Done.

danhoeflinger commented:
I squashed the commits and rebased to the current branch main.  It got screwy because of the squash and merge of the other PR and resulted in some merge conflicts.

mmichel11 reviewed (approved):
Thanks for all your work with this. The init processing logic looks sound to me and `scan_through_elements_helper` greatly improves the code quality.

I will start adding the kernels to the host code already written for oneDPL when this is merged.
